net/ipfrag/ipfrag.c

```diff
@@ -0,0 +1,1284 @@
+                              ICMPV6_EXC_FRAGTIME, 0);
+                }
+#endif
+
+              if (iob_tryadd_queue(dev->d_iob, &dev->d_fragout) == 0)
+                {
+                  netdev_iob_clear(dev);
+
+                  /* Send ICMP Time Exceeded message via dev->d_fragout
+                   * queue
+                   */
+
+                  ninfo("Send Time Exceeded ICMP%s Message to source "
+                        "host\n", node->frags->isipv4 ? "v4" : "v6");
+                  netdev_txnotify_dev(dev);
+                }
+
+              net_unlock();
+            }
+#endif
+
+          /* Remove fragments of this node */
+
+          if (node->frags != NULL)
+            {
+              FAR struct ip_fraglink_s *fraglink = node->frags;
+
+              while (fraglink)
+                {
+                  fraglink = ip_fragin_freelink(fraglink);
+                }
+            }
+
+          /* Remove node from single-list and free node memory */
+
+          ip_frag_remnode(node);
+          kmm_free(node);
+        }
+      else
+        {
+          /* Because fragment nodes have been sorted to g_assemblyhead_time
+           * according to the added time, so enter here, we can get the
+           * 'interval' of the earliest time node that has not timed out.
+           * There is no need to continue the loop here, and use time
+           * REASSEMBLY_TIMEOUT_TICKS - 'interval' as the input for the next
+           * Timer starting.
+           */
+
+          break;
+        }
+
+      entry = entrynext;
+    }
+
+  /* Be sure to start the timer, if there are nodes in the linked list */
+
+  if (sq_peek(&g_assemblyhead_time) != NULL)
+    {
+      clock_t delay = REASSEMBLY_TIMEOUT_MINIMALTICKS;
+
+      /* The interval for the next timer is REASSEMBLY_TIMEOUT_TICKS -
+       * interval, if it is less than the minimum timeout interval,
+       * fix it to REASSEMBLY_TIMEOUT_MINIMALTICKS
+       */
+
+      if (delay < REASSEMBLY_TIMEOUT_TICKS - interval)
+        {
+          delay = REASSEMBLY_TIMEOUT_TICKS - interval;
+        }
+
+      ninfo("Reschedule reassembly work queue\n");
+      wd_start(&g_wdfragtimeout, delay, ip_fragin_timerout_expiry,
+              (wdparm_t)NULL);
+    }
+  else
+    {
+      ninfo("Stop reassembly work queue\n");
+    }
+
+  nxsem_post(&g_ipfrag_mutex);
+}
+
+/****************************************************************************
+ * Name: ip_fragin_freelink
+ *
+ * Description:
+ *   Free the I/O buffer and ip_fraglink_s buffer at the head of a
+ *   ip_fraglink_s chain.
+ *
+ * Input Parameters:
+ *   fraglink - node of the lower-level linked list, it maintains
+ *              information of one fragment
+ *
+ * Returned Value:
+ *   The link to the next ip_fraglink_s buffer in the chain.
+ *
+ ****************************************************************************/
+
+static inline FAR struct ip_fraglink_s *
```

Comment:
can we avoid inline?

Reply:
> can we avoid inline?

The purpose of using inline is to save the time cost of function calls.